### PR TITLE
TL/NCCL: add gather

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Adding support for UCC_COLL_TYPE_GATHER to TL NCCL and adding gather test to ucc_test_mpi

## How ?
NCCL doesn't have native support for gather operation, implementation in TL NCCL is base on ncclSend/ncclRecv

### CODE REVIEW

File: src/components/tl/nccl/tl_nccl_coll.c [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include "utils/ucc_compiler_def.h"
 #include "utils/ucc_math.h"
 #include "utils/ucc_coll_utils.h"
+#include "utils/arch/cuda_def.h"
```
Comment (vspetrov) [inline src/components/tl/nccl/tl_nccl_coll.c:15]: copyright 2022

File: test/mpi/test_gather.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        args.dst.info.mem_type = _mt;
+        if (TEST_NO_INPLACE == inplace) {
+            args.src.info.buffer = sbuf;
+            args.src.info.count = single_rank_count;
+            args.src.info.datatype = TEST_DT;
+            args.src.info.mem_type = _mt;
+        }
+    } else {
+        args.src.info.buffer = sbuf;
+        args.src.info.count = single_rank_count;
```
Comment (vspetrov) [inline test/mpi/test_gather.cc:69]: align =

**Review by vspetrov — APPROVED** [review]

File: src/components/tl/nccl/tl_nccl_coll.c [hunk 3]
```diff
@@ -564,6 +565,85 @@ ucc_status_t ucc_tl_nccl_barrier_init(ucc_tl_nccl_task_t *task)
...
+                                    send_size, ncclChar, peer, team->nccl_comm,
+                                    stream),
+                           exit_coll, status, UCC_TL_TEAM_LIB(team));
+        }
+        NCCLCHECK_GOTO(ncclGroupEnd(), exit_coll, status,
+                       UCC_TL_TEAM_LIB(team));
+    } else {
+        NCCLCHECK_GOTO(ncclSend(src, send_size, ncclChar, args->root,
+                                team->nccl_comm, stream),
+                       exit_coll, status, UCC_TL_TEAM_LIB(team));
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/nccl/tl_nccl_coll.c:614]: that should work also, but i don't see why we should add GroupStart/GroupEnd for just one operation. According to NCCL docs it is needed in 3 cases: managing multiple GPUs from single thread, aggregating multiple operations or merging multiple send/recv.

[the remark this answers is not in the document]
Comment (kingchc) [reply src/components/tl/nccl/tl_nccl_coll.c:614]: Makes sense, thanks for the clarificaiton. I think I misunderstood the NCCL APIs.

File: src/components/tl/nccl/tl_nccl_coll.c [hunk 4]
```diff
@@ -564,6 +565,85 @@ ucc_status_t ucc_tl_nccl_barrier_init(ucc_tl_nccl_task_t *task)
...
+                            exit_coll, status);
+        }
+        NCCLCHECK_GOTO(ncclGroupStart(), exit_coll, status,
+                       UCC_TL_TEAM_LIB(team));
+        for (peer = 0; peer < size; peer++) {
+            if (peer == args->root) {
+                continue;
+            }
+            NCCLCHECK_GOTO(ncclRecv(PTR_OFFSET(dst, peer * send_size),
+                                    send_size, ncclChar, peer, team->nccl_comm,
```
Comment (kingchc) [inline src/components/tl/nccl/tl_nccl_coll.c:605]: question - is there a reason to always convert datatype to `ncclChar` as NCCL can support variety of datatypes?

**Review by bureddy — APPROVED** [review]

**Pull request merged**